mailbox/api/src/main/java/org/apache/james/mailbox/model/SimpleMailboxAnnotation.java

```diff
@@ -0,0 +1,328 @@
+                return !(value != null ? !value.equals(other.getValue()) : other.getValue() != null);
+            } else {
+                return false;
+            }
+        }
+
+        @Override
+        public int hashCode() {
+            return hash;
+        }
+
+        @Override
+        public String toString() {
+            return new StringBuilder().append(valueType).append(value).toString();
+        }
+    }
+
+    public static class SimpleMailboxAnnotationCommand implements MailboxAnnotationCommand {
+        private final MailboxAnnotationEntryKey key;
+        private final MailboxAnnotationEntryValue value;
+
+        public SimpleMailboxAnnotationCommand(MailboxAnnotationEntryKey key, MailboxAnnotationEntryValue value) {
+            this.key = key;
+            this.value = value;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (!(o instanceof SimpleMailboxAnnotationCommand)) return false;
+
+            SimpleMailboxAnnotationCommand that = (SimpleMailboxAnnotationCommand) o;
+
+            if (key != null ? !key.equals(that.key) : that.key != null) return false;
+            return !(value != null ? !value.equals(that.value) : that.value != null);
+
+        }
+
+        @Override
+        public int hashCode() {
+            int result = key != null ? key.hashCode() : 0;
+            result = 31 * result + (value != null ? value.hashCode() : 0);
+            return result;
+        }
+
+        @Override
+        public MailboxAnnotationEntryKey getEntryKey() {
+            return key;
+        }
+
+        @Override
+        public MailboxAnnotationEntryValue getEntryValue() {
+            return value;
+        }
+    }
+
+    @Override
+    public MailboxAnnotation create(MailboxAnnotationCommand annoCreate) throws UnsupportedAnnotationException {
+        return update(annoCreate);
+    }
+
+    @Override
+    public MailboxAnnotation update(MailboxAnnotationCommand annoUpdate) throws UnsupportedAnnotationException {
+        Map<MailboxAnnotationEntryKey, MailboxAnnotationEntryValue> resultEntries = new HashMap<MailboxAnnotationEntryKey, MailboxAnnotationEntryValue>(this.entries);
+        resultEntries.put(annoUpdate.getEntryKey(), annoUpdate.getEntryValue());
+        return new SimpleMailboxAnnotation(Collections.unmodifiableMap(resultEntries));
+    }
+
+    @Override
+    public MailboxAnnotation remove(MailboxAnnotationCommand annoRemove) throws UnsupportedAnnotationException {
+        Map<MailboxAnnotationEntryKey, MailboxAnnotationEntryValue> resultEntries = new HashMap<MailboxAnnotationEntryKey, MailboxAnnotationEntryValue>(this.entries);
+        if (resultEntries.containsKey(annoRemove.getEntryKey())) {
+            resultEntries.remove(annoRemove.getEntryKey());
+        }
+        return new SimpleMailboxAnnotation(Collections.unmodifiableMap(resultEntries));
+    }
+
```

Comment:
IMO update add and remove operations should be directly done against the storage, not on the list loaded on JAMES side.